lib/reversetunnel/agent.go

```diff
@@ -22,502 +22,533 @@ package reversetunnel
+	// testing.
+	clock clockwork.Clock
+	// log is an optional logger.
+	log logrus.FieldLogger
+}
+
+// checkAndSetDefaults ensures an agentConfig contains required parameters.
+func (c *agentConfig) checkAndSetDefaults() error {
+	if c.addr.IsEmpty() {
+		return trace.BadParameter("missing parameter addr")
 	}
-	if a.Client == nil {
-		return trace.BadParameter("missing parameter Client")
+	if c.sshDialer == nil {
+		return trace.BadParameter("missing parameter sshDialer")
 	}
-	if a.AccessPoint == nil {
-		return trace.BadParameter("missing parameter AccessPoint")
+	if c.transporter == nil {
+		return trace.BadParameter("missing parameter transporter")
 	}
-	if a.Signer == nil {
-		return trace.BadParameter("missing parameter Signer")
+	if c.tracker == nil {
+		return trace.BadParameter("missing parameter tracker")
 	}
-	if len(a.Username) == 0 {
-		return trace.BadParameter("missing parameter Username")
+	if c.clock == nil {
+		c.clock = clockwork.NewRealClock()
 	}
-	if a.Clock == nil {
-		a.Clock = clockwork.NewRealClock()
+	if c.log == nil {
+		c.log = logrus.New()
 	}
-	logger := a.Log
-	if a.Log == nil {
-		logger = log.StandardLogger()
+	if !c.lease.IsZero() {
+		c.log = c.log.WithField("leaseID", c.lease.ID())
 	}
-	a.Log = logger.WithFields(log.Fields{
-		trace.Component: teleport.Component(a.Component, teleport.ComponentReverseTunnelAgent),
-		trace.ComponentFields: log.Fields{
-			"target":  a.Addr.String(),
-			"leaseID": a.Lease.ID(),
-		},
-	})
-	return nil
-}
 
-// Agent is a reverse tunnel agent running as a part of teleport Proxies
-// to establish outbound reverse tunnels to remote proxies.
-//
-// There are two operation modes for agents:
-// * Standard agent attempts to establish connection
-// to any available proxy. Standard agent transitions between
-// "connecting" -> "connected states.
-// * Discovering agent attempts to establish connection to a subset
-// of remote proxies (specified in the config via DiscoverProxies parameter.)
-// Discovering agent transitions between "discovering" -> "discovered" states.
-type Agent struct {
-	sync.RWMutex
-	AgentConfig
-	log         log.FieldLogger
-	ctx         context.Context
-	cancel      context.CancelFunc
-	authMethods []ssh.AuthMethod
-	// state is the state of this agent
-	state string
-	// stateChange records last time the state was changed
-	stateChange time.Time
-	// principals is the list of principals of the server this agent
-	// is currently connected to
-	principals []string
+	c.log = c.log.WithField("target", c.addr.String())
+
+	return nil
 }
 
-// ReverseTunnelDetails contains catchable details about the reverse tunnel.
-type reverseTunnelDetails struct {
-	// TLSRoutingEnabled indicates that remote address listener supports ALPN SNI Listener and
-	// the client needs to dial the remote proxy with proper TLS ALPN protocol.
-	TLSRoutingEnabled bool
+// agent creates and manages a reverse tunnel to a remote proxy server.
+type agent struct {
```

Comment:
While a number of these fields are modified throughout the lifetime of an `agent`, it seems like only `state` is protected by a mutex. This seems like it opens the door for racy behavior if an agent is misused unintentionally, for example, calling `Start` more than once.

Reply:
The state is what protects against misuse of the agent by returning an error if an invalid state transition occurs. Only a single call to start will go through since the state will be updated from `AgentIntial`. Same goes for multiple calls to stop, once the state is set to `AgentClosed` future calls will return an error.

`doneConnecting` is also used in `Stop` to prevent accessing any fields that are set while the agent is connecting, this includes `agent.ctx, agent.drainCtx, agent.client, agent.unclaim`.

I think the same thing could be accomplished using 2 locks (1 for state, 1 for the fields set when connecting). If you think that would be more clear I can switch over to that approach. My main reasoning for choosing the `doneConnecting` channel was I didn't have to worry about the order I acquire the locks.

Reply:
Yeah I guess for most cases that should work. I'm just worried that if in the future someone decides they need to change something, for example call `connect` again outside of `Start` that could have some unintended consequences since it modifies the agent state.